### PR TITLE
generic (smart contract) event listener 

### PR DESCRIPTION
**Pull request by zelig**
The PR implements an event listener as per the appendix of https://hackmd.io/o8RGsNwZSN6IE0lkBAZCPw?both (#920)
this is a prerequisite of #922 and #934 (#896).

### CODE REVIEW

File: pkg/listener/listener.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+// - it terminates  with no error when quit channel is closed
+// - if the update function returns an error, the call returns with that error
+func (lis *EthListener) Listen(from uint64, quit chan struct{}, f func(types.Log) error) error {
+	// need to cancel context even if terminate without error with quit channel
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	// using an event channel, we subscribe to filter logs emmitted by the postage contract(s)
+	// since the last block recorded that had a relevant event
+	events := make(chan types.Log)
+	sub, err := lis.client.SubscribeFilterLogs(ctx, query(from, lis.addrs...), events)
```
Comment (ralph-pichler) [inline pkg/listener/listener.go:50]: The fromBlock used to be ignored in the past (see https://github.com/ethereum/go-ethereum/issues/15063). Not sure if this was fixed, will have to check this.

Comment (zelig) [reply pkg/listener/listener.go:50]: if that is the case we may need to read the historical logs to catch up
https://goethereumbook.org/event-read/

[the remark this answers is not in the document]
Comment (zelig) [reply pkg/listener/listener.go:50]: otherwise we cannot do subscription since there is no push notiffications, so what to do?

**Comment by acud** [comment]
closed in favor of https://github.com/ethersphere/bee/pull/1099

**Pull request closed without merging**